newrelic-agent/src/main/java/com/newrelic/agent/service/logging/LogSenderServiceImpl.java

```diff
@@ -0,0 +1,585 @@
+        if (reservoir != null) {
+            reservoir.clear();
+        }
+    }
+
+    @VisibleForTesting
+    void configureHarvestables(long reportPeriodInMillis, int maxSamplesStored) {
+        for (Harvestable h : harvestables) {
+            h.configure(reportPeriodInMillis, maxSamplesStored);
+        }
+    }
+
+    @VisibleForTesting
+    public void harvestHarvestables() {
+        for (Harvestable h : harvestables) {
+            h.harvest();
+        }
+    }
+
+    public void harvestPendingEvents() {
+        // harvest pending events
+        for (String appName : reservoirForApp.keySet()) {
+            harvestEvents(appName);
+        }
+    }
+
+    /**
+     * Store a LogEvent instance
+     * @param appName application name
+     * @param event log event
+     */
+    @Override
+    public void storeEvent(String appName, LogEvent event) {
+        if (logEventsDisabled()) {
+            return;
+        }
+
+        DistributedSamplingPriorityQueue<LogEvent> eventList = getReservoir(appName);
+        eventList.add(event);
+        Agent.LOG.finest(MessageFormat.format("Added Custom Event of type {0}", event.getType()));
+    }
+
+    /**
+     * Create and store a LogEvent instance
+     * @param appName application name
+     * @param attributes Map of attributes to create a LogEvent from
+     */
+    private void createAndStoreEvent(String appName, Map<String, ?> attributes) {
+        if (logEventsDisabled()) {
+            return;
+        }
+
+        DistributedSamplingPriorityQueue<LogEvent> eventList = getReservoir(appName);
+        eventList.add(createValidatedEvent(attributes));
+        Agent.LOG.finest(MessageFormat.format("Added event of type {0}", LOG_EVENT_TYPE));
+    }
+
+    /**
+     * Check if LogEvents are disabled
+     *
+     * @return true if they are disabled, false if they are enabled
+     */
+    private boolean logEventsDisabled() {
+        if (!forwardingEnabled) {
+            if (ServiceFactory.getConfigService().getDefaultAgentConfig().isHighSecurity()) {
+                Agent.LOG.log(Level.FINER, "Event of type {0} not collected due to high security mode being enabled.", LOG_EVENT_TYPE);
+            } else {
+                Agent.LOG.log(Level.FINER, "Event of type {0} not collected. application_logging.forwarding not enabled.", LOG_EVENT_TYPE);
+            }
+            Agent.LOG.log(Level.FINER, "Event of type {0} not collected. application_logging.forwarding not enabled.", LOG_EVENT_TYPE);
+            return true; // LogEvents are disabled
+        }
+        return false; // LogEvents are enabled
+    }
+
+    /**
+     * Get the LogEvent reservoir
+     *
+     * @param appName app name
+     * @return Queue of LogEvent instances
+     */
+    @VisibleForTesting
+    public DistributedSamplingPriorityQueue<LogEvent> getReservoir(String appName) {
+        DistributedSamplingPriorityQueue<LogEvent> result = reservoirForApp.get(appName);
+        while (result == null) {
```

Comment:
So is there any chance this will never be non-null?  I'm guessing not but at a glance it scares me a bit.

Reply:
I suppose so but it shouldn't matter should it as we just return the queue in the case that it is non-null?

Reply:
sorry...I mistated...I was wondering if we got into the loop and then possibly never got out

Reply:
I don't think that should be possible. If it gets into the while loop then it adds an empty queue associated with the `appName` argument and assigns that to `result`. At that point the `result` queue would no longer be `null` and it should exit the while loop.

We have the same logic in other services and it's never been an issue. https://github.com/newrelic/newrelic-java-agent/blob/3b934f3ccfbd15437a2f85ef705979148ab5a293/newrelic-agent/src/main/java/com/newrelic/agent/service/analytics/InsightsServiceImpl.java#L252-L261